### PR TITLE
sync: ensure Mutex, RwLock, and Semaphore futures are Send + Sync

### PR DESCRIPTION
**Pull request by hawkw**
## Motivation

Previously, the `Mutex::lock`, `RwLock::{read, write}`, and
`Semaphore::acquire` futures in `tokio::sync` implemented `Send + Sync`
automatically. This was by virtue of being implemented using a `poll_fn`
that only closed over `Send + Sync` types. However, this broke in
PR #2325, which rewrote those types using the new `batch_semaphore`.
Now, they await an `Acquire` future, which contains a `Waiter`, which
internally contains an `UnsafeCell`, and thus does not implement `Sync`.

Since removing previously implemented traits breaks existing code, this
inadvertantly caused a breaking change. There were tests ensuring that
the `Mutex`, `RwLock`, and `Semaphore` types themselves were `Send +
Sync`, but no tests that the _futures they return_ implemented those
traits.

## Solution

I've fixed this by adding an explicit impl of `Sync` for the
`batch_semaphore::Acquire` future. Since the `Waiter` type held by this
struct is only accessed when borrowed mutably, it is safe for it to
implement `Sync`.

Additionally, I've added to the bounds checks for the effected
`tokio::sync` types to ensure that returned futures continue to
implement `Send + Sync` in the future.

Fixes: #2373

Signed-off-by: Eliza Weisman <eliza@buoyant.io>

### CODE REVIEW

**Comment by hawkw** [comment]
> Does this need a bump release? if so we should prepare that along side this and we can get it out asap

I think we probably should do a version bump — @carllerche, what do you think?

I think that should be staged in a separate PR though.

**Comment by LucioFranco** [comment]
@hawkw cool, if you can open a bump Pr I can do the rest.

**Comment by carllerche** [comment]
We should push a patch release. You can include the version bump / changelog here too.

**Comment by hawkw** [comment]
@carllerche okay, added the version bump in 89fcffb62682a5bb703f88163fc7c8aa5c07d75f

File: tokio/Cargo.toml [hunk 1]
```diff
@@ -8,12 +8,12 @@ name = "tokio"
...
 # - Update CHANGELOG.md.
 # - Create "v0.2.x" git tag.
-version = "0.2.15"
+version = "0.2.16"
 edition = "2018"
 authors = ["Tokio Contributors <team@tokio.rs>"]
 license = "MIT"
 readme = "README.md"
-documentation = "https://docs.rs/tokio/0.2.15/tokio/"
+documentation = "https://docs.rs/tokio/0.1.16/tokio/"
```
Comment (carllerche) [inline tokio/Cargo.toml:16]:
```suggestion
documentation = "https://docs.rs/tokio/0.2.16/tokio/"
```

File: tokio/src/lib.rs [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#![doc(html_root_url = "https://docs.rs/tokio/0.2.15")]
+#![doc(html_root_url = "https://docs.rs/tokio/0.1.16")]
```
Comment (carllerche) [inline tokio/src/lib.rs:1]:
```suggestion
#![doc(html_root_url = "https://docs.rs/tokio/0.2.16")]
```

**Comment by hawkw** [comment]
Whoops, thanks for catching my typo @carllerche!

**Comment by carllerche** [comment]
It would be nice if azure windows CI stopped randomly not reporting back....

**Comment by hawkw** [comment]
@carllerche looks like it's good now?

**Pull request merged**